NumericField: restore preventDefault to true on up/down arrows

IIARROWS opened the pull request:
Fix issue #1543, onkeydown:preventDefault on the input tag wasn't set to true when pressing up/down arrows, causing double increment: the code and the browser.

This PR is draft because I noticed that there are some quirks on the restore of "preventDefault" on other inputs.

IIARROWS commented:
I encountered another issue with the fix, and I think it involves the changes made for MudAutocomplete so I need support on this.

Basically it doesn't double anymore, but other inputs are disabled until focus is lost or enter is pressed (in Immediate mode, in Normal or Debounced just for the key pressed immediately after the arrow, the next one works as intended).
It's still an improvement, if you think it can go ahead in this state you merge it.

henon commented:
Can you explain in detail how to reproduce the problem? I just tried your branch and didn't see anything unusual.

IIARROWS commented:
Try run the docs app, under in the numeric field page you can easily reproduce it in the "Normal vs Immediate vs Bounced" section and see the difference between the 3 modes.

1. Use the up/down arrows, the number should change correctly
2. Pressing any other key (like TAB, a number or the right/left arrow to move the cursor), produces no visible effect (Enter commits the value on all 3 modes)
3. If Normal or Debounced, the next keystroke works as intended (for example, Tab changes focus on the next element)
If Immediate, the next keystroke doesn't work, unless the previous one was Enter)

I'm pretty sure Step 2 is an effect on the implementation of KeyDownPreventDefault
I think Enter works on Immediate because of the changes made by @porkopek on `MudBaseInput` that added `ShouldRenderBeForced` that works on Enter, ArrowUp and ArrowDown.

henon commented:
This is probably an effect of my original fix for Autocomplete swallowing characters (happened also on Input and Textfield when you typed very fast). I suppressed rendering as long as the user is typing (by setting _isFocused true in keydown) to allow the input to handle immediate user input without being constantly interruped by unnecessary renders which would overwrite the user input with an old value of the input's text. 

This prevention of rendering the input while focused (focus state is tracked in `_isFocused `) requires that in certain situations you need to force a render update. This is what Porkopek did for autocomplete, namely to forceRender it when the Arrow up/down keys are pressed. What you can do is either set _shouldRenderBeForced to true when you want to force the next render followed by certain keypresses or you set _isFocused to false when something happens that indicates that the user is not having focus on the input any more.

IIARROWS commented:
Yeah, I didn't investigate that far.
I tried an hack to force `_referenceElementl._shouldRenderBeForced` from MudNumberInput in case `_keyDownPreventDefault` was true before it gets overridden by the public parameter `KeyDownPreventDefault`* but it didn't work, probably because I was doing something wrong at the time.

* that is for the developer in case they want to override che component behaviour, forcing

henon commented:
I also tried to set _shouldRenderBeForced to true in `InterceptArrowKey` but no difference. I think you are right that the preventdefault is the problem.

henon commented:
The problem vanishes when you uncomment the `_keyDownPreventDefault=true;` lines in `InterceptArrowKey`

henon commented:
I think I found a fix, please try it.

henon commented:
Lol, I thought that the lines `_keyDownPreventDefault = true;` are necessary to fix the issue of incrementing twice. After testing again without them the NumericField works just fine too with a Step of 0.1. Could it be that switching from KeyPress to KeyDown already fixed it and that they are just totally unnecessary?

henon commented:
I don't get it. I checked out current dev and the value does not double increment with a Step of 0.1 and using the arrow keys.

henon commented:
Aha, it happens only on WASM

henon commented:
Reverted. It did fix the problem caused by preventDefault but it also caused double increment again.

henon commented:
Interesting: when you do this then both issues are gone:

```c#
        protected void InterceptArrowKey(KeyboardEventArgs obj)
        {
            if (Disabled || ReadOnly)
                return;

            if (obj.Type == "keydown")//KeyDown or repeat, blazor never fires InvokeKeyPress
            {
                if (obj.Key == "ArrowUp")
                {
                    //_keyDownPreventDefault = true;
                    //_ = Increment();
                    return;
                }
                else if (obj.Key == "ArrowDown")
                {
                    //_keyDownPreventDefault = true;
                    //_ = Decrement();
                    return;
                }
            }
            _keyDownPreventDefault = KeyDownPreventDefault;
            OnKeyDown.InvokeAsync(obj).AndForget();
        }
```

But what other issues does it cause? Overflow? What is the purpose of this manual increment?

IIARROWS commented:
No, they are “not”.

From the method summary:

> Overrides KeyDown event, intercepts Arrow Up/Down and uses them to add/substract the value manually by the step value.

> Relying on the browser mean the steps are each integer multiple from <see cref="Min"/> up until <see cref="Max"/>.

> This align the behaviour with the spinner buttons.



If you go to the test I posted on Discord, you can see that with a step of 1.3, 3.142 becomes 4.1 instead of 4.442 because the browser goes in steps of 1.3 from the min of -5.

That’s unlike any other application I’ve ever seen, I have no idea why the W3C decided that way.

henon commented:
ok, got it. I removed all my commits from your PR via forcePush, they were useless. But I am onto something ...

henon commented:
This time I nailed it I think. Here is what I found out:

1) The duplication of increment/decrement happened only on WASM
2) The fix for that duplication then caused the next keystroke to be swallowed on Server
3) MudInputs's render-protection while typing in Immediate mode caused all subsequent input to be swallowed after using the up- and down keyboard keys until you hit Enter on both WASM and Server

To Fix 2) I added a service called `RuntimeLocation` which can be used to distinguish between WASM and server. So the preventDefault hack is only done on WASM now.

To Fix 3) I made the MudInput render-protection optional and disabled it for the input of NumericField.

IIARROWS commented:
But now on server, the arrows work on browser logic.

IIARROWS commented:
Sure, but in that case the double input comes back 😊

henon commented:
> But now on server, the arrows work on browser logic.

You mean the Arrow keyboard keys? I single stepped and saw the InterceptArrowKey function called both on server and on wasm.

> Sure, but in that case the double input comes back 😊

What do you mean exactly?

IIARROWS commented:
I sent the second message at 14:17 UTC by email....

IIARROWS commented:
> > But now on server, the arrows work on browser logic.
> 
> You mean the Arrow keyboard keys? I single stepped and saw the InterceptArrowKey function called both on server and on wasm.

OK, but the value is not replaced.
I'm testing like this, changing the doc example "DebouncedNumericExcample", changing the variable from int to double,
if you change the value to 0.5, and press up arrow you would expect 1.5, but you get 1 instead, and then 2.

IIARROWS commented:
As discussed privately with henon, removed the RuntimeLocation from service to an utility class, to be expanded in the future.
Implemented a new interceptor OnKeyUp to froce reset the value of `_keyDownPreventDefault`. This fixes the problem on WASM.

**There is no double-step or non-responsive keyboard bug with this version.**
The only issue is in Server rendering, on keyboard arrow up/down the behaviour is the browser default.

The browser first changes the value, then server gets called, and Blazor engine doesn't replace the value with the one given by the server.
I'm not even sure it's possible to solve in server mode, I'll investigate further and open a question on MS Asp.Net GitHub as a last resort.

As for now, we have this difference in behaviour in WASM vs Server, we should define if it's acceptable for MudBlazor to have that and in case document it.

henon commented:
ok, let's document the behavior, no need to waste hours on that little quirk

henon commented:
actually like this it could be accurately be incremented also on server, I tested this and it works in principle. 

```c#
                    var value = Value;
                    await Task.Delay(1);
                    Value = value;
                    await Increment();
``` 

the only drawback is that it would cause a render update on every key press. that in itself is not a problem but if you hold down the key and keep it down somehow blazor reduces the render frequency and it will slow down.

henon commented:
This is the best of both worlds, it is accurate on server and the issue with slowing down happens only on WASM so by doing it only on server everything is fine

```c#
                if (obj.Key == "ArrowUp")
                {
                    _keyDownPreventDefault = true;
                    if (RuntimeLocation.IsServerSide)
                    {
                        var value = Value;
                        await Task.Delay(1);
                        Value = value;
                    }
                    await Increment();
                    return;
                }
                else if (obj.Key == "ArrowDown")
                {
                    _keyDownPreventDefault = true;
                    if (RuntimeLocation.IsServerSide)
                    {
                        var value = Value;
                        await Task.Delay(1);
                        Value = value;
                    }
                    await Decrement();
                    return;
                }
```

henon commented:
Thanks @IIARROWS !
